telemetry/Tracks/class-tracks-event.php

```diff
@@ -0,0 +1,275 @@
+	}
+
+	/**
+	 * Returns the event's data.
+	 *
+	 * @return stdClass|WP_Error Event object if the event was created successfully, WP_Error otherwise.
+	 */
+	public function get_data() {
+		return $this->data;
+	}
+
+	/**
+	 * Returns whether the event can be recorded.
+	 *
+	 * @return bool|WP_Error True if the event is recordable.
+	 *                       False if the event is not recordable.
+	 *                       WP_Error if the event is generating an error.
+	 */
+	public function is_recordable() {
+		if ( ! Telemetry_System::is_wpadmin_telemetry_allowed() ) {
+			return false;
+		}
+
+		// Don't record events during unit tests and CI runs.
+		if ( 'wptests_capabilities' === wp_get_current_user()->cap_key ) {
+			return false;
+		}
+
+		if ( is_wp_error( $this->data ) ) {
+			return $this->data;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Processes the event's properties to get them ready for validation.
+	 *
+	 * @param string $event_name The event's name.
+	 * @param array<string, mixed>|array<empty> $event_properties Any event properties to be processed.
+	 * @return stdClass The resulting event object with processed properties.
+	 */
+	protected static function process_properties(
+		string $event_name,
+		array $event_properties
+	): stdClass {
+		$event = (object) static::sanitize_properties_array( $event_properties );
+		$event = static::set_user_properties( $event );
+
+		// Set event name.
+		$event->_en = preg_replace(
+			'/^(?:' . static::EVENT_NAME_PREFIX . ')?(.*)/',
+			static::EVENT_NAME_PREFIX . '\1',
+			$event_name
+		) ?? '';
+
+		// Set event timestamp.
+		if ( ! isset( $event->_ts ) ) {
+			$event->_ts = static::get_timestamp();
+		}
+
+		// Remove non-routable IPs to prevent record from being discarded.
+		if ( property_exists( $event, '_via_ip' ) &&
+			1 === preg_match( '/^192\.168|^10\./', $event->_via_ip ) ) {
+			unset( $event->_via_ip );
+		}
+
+		// Set VIP environment if it exists.
+		if ( defined( 'VIP_GO_APP_ENVIRONMENT' ) ) {
+			$app_environment = constant( 'VIP_GO_APP_ENVIRONMENT' );
+			if ( is_string( $app_environment ) && '' !== $app_environment ) {
+				$event->vipgo_env = $app_environment;
+			}
+		}
+
+		return $event;
+	}
+
+	/**
+	 * Sets the Tracks User ID and User ID Type depending on the current
+	 * environment.
+	 *
+	 * @param stdClass $event The event to annotate with identity information.
+	 * @return stdClass The new event object including identity information.
+	 */
+	protected static function set_user_properties( stdClass $event ): stdClass {
+		$wp_user_id = get_current_user_id();
+
+		// Only track logged-in users.
+		if ( 0 === $wp_user_id ) {
+			return $event;
+		}
+
+		// Users in the VIP environment.
+		if ( defined( 'VIP_GO_APP_ID' ) ) {
+			$app_id = constant( 'VIP_GO_APP_ID' );
+			if ( is_integer( $app_id ) && 0 < $app_id ) {
+				$event->_ui = $app_id . '_' . $wp_user_id;
+				$event->_ut = 'vip_go_app_wp_user';
+
+				return $event;
+			}
```

Comment:
We need to use an identifier that is pseudo-anonymized but still allows tracking the user across different environments (e.g. if I'm working on wpvip.com and docs.wpvip.com, I should not be counted as two different users).

Maybe a hash of the email?

```
$salt = constant( 'VIP_TELEMETRY_SALT' ); // this would be private but globally shared across the platform
$tracks_user_id = hash_hmac( 'sha256', $user->user_email, $salt );
```

Reply:
Added this in 89b0248fc7bd0dc74b6995ef8a64f4731551d916